### DISCUSSION TITLE
Signal/event/observer pattern

### DISCUSSION THREAD

**Discussion by musjj** [hajimehoshi/ebiten#2787] [Q&A] [answered]
Does anyone have any good examples on how you'd implement a signal/event/observer pattern in Ebiten and Go?

Examples of such patterns:

https://docs.godotengine.org/en/stable/getting_started/step_by_step/signals.html

https://gameprogrammingpatterns.com/observer.html

**Comment by SolarLune** [comment, marked as answer]
You could do this with message passing - a message can be anything and fulfill a certain interface. You register listeners with some global object, and then send messages to those registered listeners. Those listeners can be any kind of object. 

Here's a repo that does that: https://github.com/SolarLune/messages

You could also just do this with interfaces directly, as objects can be checked to see if they implement an interface, which would indicate they subscribe to the desired message type.

**Reply by musjj** [reply]
Thank you, that's a really simple & smart way to do it! It's also type-safe which is really nice!
